crates/codegen/schema/src/validation/ast/visitors.rs

```diff
@@ -52,9 +58,37 @@ pub trait Visitor {
         return VisitorResponse::StepIn;
     }
 
-    fn visit_expression(
+    fn visit_scanner(
+        &mut self,
+        _scanner: &Rc<Scanner>,
+        _reporter: &mut Reporter,
+    ) -> VisitorResponse {
+        return VisitorResponse::StepIn;
+    }
+
+    fn visit_parser(&mut self, _parser: &Rc<Parser>, _reporter: &mut Reporter) -> VisitorResponse {
+        return VisitorResponse::StepIn;
+    }
+
+    fn visit_precedence_parser(
         &mut self,
-        _expression: &ExpressionRef,
+        _precedence_parser: &Rc<PrecedenceParser>,
+        _reporter: &mut Reporter,
+    ) -> VisitorResponse {
+        return VisitorResponse::StepIn;
+    }
+
+    fn visit_operator_definition(
```

Comment:
I wonder if you foresee a validation rule making use of `visit_operator_definition()` or `visit_primary_expression_reference()` without the containing `visit_precedence_parser()` context/data?